File: co_testing/unit_tests/ADPCM_Dec/test_ADPCMdec.c
/*#########################################################################
#  P & D Embedded Systems and Multimedia [H09M0a]
#  Subband-Coding
#
#  Mohit Dandekar
#  John O'Callaghan
#########################################################################
#
#  Test Wrapper for unit test ADPCM Decoder
#
#  + Reads testvectors: data_in,  expected output
#  + Calls C-kernel 'ADPCM_dec_process'
#  + compares DUT output with REF
#######################################################################*/

#include <stdio.h>
#include <stdlib.h>

#include "ADPCM.h"

#define P 738*4
#define N P*BLOCK_SIZE/4

ADPCM_dec_t ADPCM_dec;

adpcm_data_t data_in[P*BLOCK_SIZE];
aud_data_t data_out_ref[P*BLOCK_SIZE];
aud_data_t data_out_dut[P*BLOCK_SIZE];

sint16 mu;
sint8 nbit;
sint16 blk_len;

aud_data_t temp_buffer[BLOCK_SIZE];

int main(void)
{
  sint32 ret = 0;
  sint32 i,j;

  /*Read Testvector data*/
  printf("Reading Testvectors\n");
  FILE *fp;
  fp = fopen("./test_vectors/data_in.txt","r");
  
  for(i=0;i<BLOCK_SIZE*P;++i)
  {
    int templ,tempr;
    fscanf(fp,"%d,%d",&templ,&tempr);
    data_in[i].l = templ;
    data_in[i].r = tempr;
  }
  fclose(fp);

  fp = fopen("./test_vectors/adpcm_expected.txt","r");
  for(i=0;i<BLOCK_SIZE*P;++i)
  {
    int templ,tempr;
    fscanf(fp,"%d,%d",&templ,&tempr);
    data_out_ref[i].l = templ;
    data_out_dut[i].l = 0;
    data_out_ref[i].r = tempr;
    data_out_dut[i].r = 0;
  }
  fclose(fp);

  mu = 13222; // 0.4035
  nbit = 5;
  blk_len = BLOCK_SIZE/4;

  /*Initialization*/
  printf("Initialize: C kernel under Test\n");
  ADPCM_dec_init( &ADPCM_dec,
                  temp_buffer,
                  mu,
                  nbit,
                  blk_len);

  /*Process Call*/
  printf("Process Call: C kernel under Test\n");
  for(i=0;i<P;++i)
  {

    /*Write  Input Buffer of the encoder*/
   for(j=0;j<blk_len;++j)
    {
      ADPCM_dec.data_in_buffer[j] = data_in[i*blk_len + j];
    }

    /*ADPCM dec process call*/
    ADPCM_dec_process(&ADPCM_dec);

    /*Read from temporary out buffer to main memory*/
    for(j=0;j<blk_len;++j)
    {
       data_out_dut[i*blk_len + j] = temp_buffer[j];
    }
  }

  /* Compare Output vs Ref*/
  ret = 0;
  printf("Checking output\n");
  fp = fopen("./test_vectors/output.txt","w");
  for(i=0;i<P*blk_len;++i)
  {
    fprintf(fp,"%d,%d\n",data_out_dut[i].l,data_out_dut[i].r);
    /*Allowing no errors*/
    if(abs(data_out_ref[i].l - data_out_dut[i].l)>0 || abs(data_out_ref[i].r - data_out_dut[i].r)>0)
    {
      printf("ERROR : indx = %8d expected = %8d,%8d  output %8d %8d\n",i,data_out_ref[i].l,
                                                                          data_out_ref[i].r,
                                                                          data_out_dut[i].l,
                                                                          data_out_dut[i].r);
      ret += 1;
    }
  }
  fclose(fp);
  if(ret)
  {
         printf("                          TEST FAIL                               \n");
         printf("  total errors: %d of %d \n",ret,P*blk_len                           );
         printf("##################################################################\n");

  }else{
         printf("                          TEST PASS                               \n");
         printf("##################################################################\n");
       }

  return 0;
}

File: co_testing/unit_tests/ADPCM_Dec/test_ADPCMdec.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  P & D Embedded Systems and Multimedia [H09M0a] 2015-2016
%  Subband-Coding 
%
%  Mohit Dandekar
%  John O'Callaghan
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Unit Test for ADPCM Decoder
%
%  + loads input WAV
%  + Processes using ADPCM_coder MATLAB kernel to produce output
%  + Writes the testvectors: input_data, adpcm expected output
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

clear all;
close all;
%clc;
if(exist('OCTAVE_VERSION', 'builtin') ~= 0)
  pkg load signal;
end
addpath('../../../MATLAB/src');
addpath('../../../MATLAB/utils');

f2 = fopen('./test_vectors/test_name.txt','w');
fprintf(f2,'ADPCM_Decoder');
fclose(f2);
fprintf('#################################################################\n');
printtime();
fprintf('     Unit Testcase for ADPCM Decoder                  \n\n');

load('../../../training_data/words_m')

%yl = convert_to_FXPT((rand(N,1)-0.5),16,15);
%yr = convert_to_FXPT((rand(N,1)-0.5),16,15);
xl = x(1:2:length(x));
yl = resample(xl,8000,16000); % includes anti-alising
yl = yl(1:(floor(length(yl)/10/24))*24);
xr = x(2:2:length(x));
yr = resample(xr,8000,16000); % includes anti-alising
yr = yr(1:(floor(length(yr)/10/24))*24);

yl = convert_to_FXPT(yl,16,15);
yr = convert_to_FXPT(yr,16,15);

L = 24/4;
N = length(yl);
P = N/L;
fprintf('Number of Input blocks = %d  Block Size = %d\n',P,L);

fs = fs/2;

mode = 0;

mu1 = 0.4035;
mu1 = convert_to_FXPT(mu1,16,15); 

nbit_1 = 5;

coder_history_l = [16384 0 0];
coder_history_r = [16384 0 0];
decoder_history_l = [16384 0 0];
decoder_history_r = [16384 0 0];

start_indx = [1]; end_indx = floor([length(yl)]/P);
incr = end_indx;

fprintf('ADPCM Decoder\n');fprintf('Block : [%d]',P);

for i=1:P
  fprintf('%8d',i);
  if(exist('OCTAVE_VERSION', 'builtin') ~= 0)
    fflush(1);
  end    

  [z000l(start_indx(1):end_indx(1)), coder_history_l] = ADPCM_coder(yl(start_indx(1):end_indx(1)), mu1, nbit_1, coder_history_l, mode);
  [z000r(start_indx(1):end_indx(1)), coder_history_r] = ADPCM_coder(yr(start_indx(1):end_indx(1)), mu1, nbit_1, coder_history_r, mode);

  [youtl(start_indx(1):end_indx(1)), decoder_history_l(1,:)] = ADPCM_decoder(z000l(start_indx(1):end_indx(1)), mu1, nbit_1, decoder_history_l(1,:), mode);
  [youtr(start_indx(1):end_indx(1)), decoder_history_r(1,:)] = ADPCM_decoder(z000r(start_indx(1):end_indx(1)), mu1, nbit_1, decoder_history_r(1,:), mode);


  fprintf('\b\b\b\b\b\b\b\b');
  start_indx = start_indx + incr;
  end_indx = end_indx + incr;
end


fprintf('\nWriting Testvectors\n');


f2 = fopen('./test_vectors/data_in.txt','w');
for i=1:N
  fprintf(f2,'%d,%d\n',z000l(i),z000r(i));
end
fclose(f2);

f1 = fopen('./test_vectors/adpcm_expected.txt','w');
for i=1:N
  fprintf(f1,'%d,%d\n',youtl(i),youtr(i));
end
fclose(f1);


